### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by stephchen**
Description
-----------

refactors launch code so that launch and launch-add largely share the same path for clarity & testing reasons:
- cli generates a run spec conforming to the json spec in gorilla. spec is sent to server in launch-add case
- project generated off spec
- project generation and local fetching are separated

Testing
-------

tests updated

### CODE REVIEW

File: wandb/sdk/launch/utils.py [hunk 1]
```diff
@@ -50,38 +51,80 @@ def _is_wandb_local_uri(uri: str):
...
 ):
-    parameters = parameters or {}
-    experiment_name = experiment_name
-    project = _project_spec.Project(
+    # override base config (if supplied) with supplied args
+    run_spec = launch_config if launch_config is not None else {}
+    run_spec["uri"] = uri
+    run_spec["entity"] = wandb_entity
+    run_spec["project"] = wandb_project
+    run_spec["name"] = experiment_name
```
Comment (KyleGoyette) [inline wandb/sdk/launch/utils.py:82]: won't this automatically overwrite the values in the run_spec even if the args aren't given?

File: wandb/sdk/launch/docker.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def generate_docker_image(project: _project_spec.Project, entry_cmd):
     cmd: Sequence[str] = [
         "jupyter-repo2docker",
         "--no-run",
-        "--user-id={}".format(project.user_id),
+        "--user-id={}".format(project.docker_user_id),
```
Comment (KyleGoyette) [inline wandb/sdk/launch/docker.py:55]: nice this is clearer

**Review by KyleGoyette — COMMENTED** [review]
One question, also looks like formatting and flake8 issues need to be fixed.

There's also a new issue building the launch tests on CI, did you run the launch tests locally? I'll file a ticket to fix the CI issue.

File: wandb/sdk/launch/utils.py (hunk 1 above)
Comment (stephchen) [reply wandb/sdk/launch/utils.py:82]: oh yeah you're right idk why i did that

**Pull request merged**